core/common/src/main/java/alluxio/util/JvmPauseMonitor.java

```diff
@@ -0,0 +1,195 @@
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.util;
+
+import alluxio.PropertyKey;
+import alluxio.Configuration;
+
+import com.google.common.base.Joiner;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import com.google.common.base.Stopwatch;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import java.lang.management.GarbageCollectorMXBean;
+import java.lang.management.ManagementFactory;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Class to monitor JVM with a daemon thread, the thread sleep period of time
+ * and get the true time the sleep takes. If it is longer than it should be, the
+ * JVM has paused processing. Then log it into different level.
+ */
+public final class JvmPauseMonitor {
+  private static final Log LOG = LogFactory.getLog(
+      JvmPauseMonitor.class);
+
+  /** The time to sleep. */
+  private final long mGcSleepIntervalMs;
+
+  /** Extra sleep time longer than this threshold, log WARN. */
+  private final long mWarnThresholdMs;
+
+  /** Extra sleep time longer than this threshold, log INFO. */
+  private final long mInfoThresholdMs;
+
+  /** Times extra sleep time exceed WARN. */
+  private long mExceedWarnTimes = 0;
+  /** Times extra sleep time exceed INFO. */
+  private long mExceedInfoTimes = 0;
+  /** Total extra sleep time. */
+  private long mTotalExtraTime = 0;
+
+  private Thread mJvmMonitorThread;
+  private volatile boolean mThreadStarted = true;
+
+  /**
+   * Constructs JvmPauseMonitor.
+   */
+  public JvmPauseMonitor() {
+    mGcSleepIntervalMs = Configuration.getLong(PropertyKey.JVM_MONITOR_SLEEP_INTERVAL_MS);
+    mWarnThresholdMs = Configuration.getLong(PropertyKey.JVM_MONITOR_WARN_THRESHOLD_MS);
+    mInfoThresholdMs = Configuration.getLong(PropertyKey.JVM_MONITOR_INFO_THRESHOLD_MS);
+  }
+
+  /**
+   * Starts jvm monitor thread.
+   */
+  public void start() {
+    Preconditions.checkState(mJvmMonitorThread == null,
+        "JVM monitor thread already started");
+    mJvmMonitorThread = new Daemon(new Monitor());
+    mJvmMonitorThread.start();
+  }
+
+  /**
+   * Stops jvm monitor.
+   */
+  public void stop() {
+    mThreadStarted = false;
+    mJvmMonitorThread.interrupt();
+    try {
+      mJvmMonitorThread.join();
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+    }
+    reset();
+  }
+
+  /**
+   * Resets value of mJvmMonitorThread and mThreadStarted.
+   * @return reseted mThreadStarted
+   */
+  public JvmPauseMonitor reset() {
+    mJvmMonitorThread = null;
+    mThreadStarted = true;
```

Comment:
the thread is not started, since it's set to null above,right?

Reply:
In fact `mThreadStarted ` seems not so useful given `mJvmMonitorThread ` is used for checking if the thread is null.